pkg/controller/logstash/config.go

```diff
@@ -85,3 +116,158 @@ func defaultConfig() *settings.CanonicalConfig {
+func checkTLSConfig(config *configs.APIServer, useTLS bool) error {
+	svcUseTLS := strconv.FormatBool(useTLS)
+	sslEnabled := config.SSLEnabled
+	if (svcUseTLS == sslEnabled) || (!useTLS && sslEnabled == "") {
+		return nil
+	}
+
+	return fmt.Errorf("API Service `spec.services.tls.selfSignedCertificate.disabled` is set to `%t`, but logstash config `api.ssl.enabled` is set to `%s`", !useTLS, sslEnabled)
+}
+
+// resolveAPIServerConfig gives ExpectedAPIServer with the resolved ${VAR} value
+func resolveAPIServerConfig(cfg *settings.CanonicalConfig, params Params) (*configs.APIServer, error) {
+	config := baseAPIServer(cfg)
+
+	if unresolvedConfig := patchWithPatternValue(config); len(unresolvedConfig) > 0 {
+		combinedMap, err := getKeystoreEnvKeyValue(params)
+		if err != nil {
+			return nil, err
+		}
+
+		resolveConfigValue(unresolvedConfig, combinedMap)
+	}
+
+	return config, nil
+}
+
+// baseAPIServer gives api.* configs with the default value
+func baseAPIServer(cfg *settings.CanonicalConfig) *configs.APIServer {
+	enabled, _ := cfg.String("api.ssl.enabled")
+	keystorePassword, _ := cfg.String("api.ssl.keystore.password")
+	authType, _ := cfg.String("api.auth.type")
+	username, _ := cfg.String("api.auth.basic.username")
+	pw, _ := cfg.String("api.auth.basic.password")
+
+	return &configs.APIServer{
+		SSLEnabled:       enabled,
+		KeystorePassword: keystorePassword,
+		AuthType:         authType,
+		Username:         username,
+		Password:         pw,
+	}
+}
+
+// patchWithPatternValue matches the pattern ${VAR:default_value} against config and assign the default value
+// It gives a map of string (VAR) and config pointer that need to further resolve
+// VAR is the variable name that expect to be defined in Keystore or Env
+//
+//	The variable name can consist of digit, underscores and letters
+//	The default value is optional, ${VAR:} and ${VAR} are valid.
+//
+// config is updated with the default value
+func patchWithPatternValue(config *configs.APIServer) map[string]*string {
+	data := make(map[string]*string)
+
+	pattern := `^\${([a-zA-Z0-9_]+)(?::(.*?))?}$`
+	regex := regexp.MustCompile(pattern)
+
+	for _, configKey := range []*string{&config.SSLEnabled, &config.KeystorePassword, &config.AuthType, &config.Username, &config.Password} {
+		if match := regex.FindStringSubmatch(*configKey); match != nil {
+			key := match[1]
+			defaultValue := match[2]
+			*configKey = defaultValue
+			data[key] = configKey
+		}
+	}
+
+	return data
+}
+
+// getKeystoreEnvKeyValue gives a map that consolidate all key value pairs from user defined environment variables
+// and Keystore from SecureSettings. If the same key defined in both places, keystore takes the precedence.
+func getKeystoreEnvKeyValue(params Params) (map[string]string, error) {
+	data := make(map[string]string)
+	c := getLogstashContainer(params.Logstash.Spec.PodTemplate.Spec.Containers)
```

Comment:
The return value from `getLogstashContainer` should be checked - the value will be `nil` with a CRD definition not containing an explicit container definition in the `podTemplate` spec, which will cause a crash in subsequent checks of the `Env` and `EnvFrom`, etc below.